File: pulse-kotlin/src/commonMain/kotlin/com/fluxchess/pulse/kotlin/engine/Piece.kt
/*
 * Copyright 2013-2024 Phokham Nonava
 *
 * Use of this source code is governed by the MIT license that can be
 * found in the LICENSE file.
 */

package com.fluxchess.pulse.kotlin.engine

typealias Piece = Int

const val WHITE_PAWN: Piece = 0
const val WHITE_KNIGHT: Piece = 1
const val WHITE_BISHOP: Piece = 2
const val WHITE_ROOK: Piece = 3
const val WHITE_QUEEN: Piece = 4
const val WHITE_KING: Piece = 5
const val BLACK_PAWN: Piece = 6
const val BLACK_KNIGHT: Piece = 7
const val BLACK_BISHOP: Piece = 8
const val BLACK_ROOK: Piece = 9
const val BLACK_QUEEN: Piece = 10
const val BLACK_KING: Piece = 11

const val NO_PIECE: Piece = 12

val pieces = intArrayOf(
	WHITE_PAWN, WHITE_KNIGHT, WHITE_BISHOP, WHITE_ROOK, WHITE_QUEEN, WHITE_KING,
	BLACK_PAWN, BLACK_KNIGHT, BLACK_BISHOP, BLACK_ROOK, BLACK_QUEEN, BLACK_KING,
)

fun isValidPiece(piece: Piece): Boolean = when (piece) {
	WHITE_PAWN, WHITE_KNIGHT, WHITE_BISHOP, WHITE_ROOK, WHITE_QUEEN, WHITE_KING,
	BLACK_PAWN, BLACK_KNIGHT, BLACK_BISHOP, BLACK_ROOK, BLACK_QUEEN, BLACK_KING,
		-> true

	else -> false
}

fun pieceOf(color: Color, pieceType: PieceType): Piece = when (color) {
	WHITE -> when (pieceType) {
		PAWN -> WHITE_PAWN
		KNIGHT -> WHITE_KNIGHT
		BISHOP -> WHITE_BISHOP
		ROOK -> WHITE_ROOK
		QUEEN -> WHITE_QUEEN
		KING -> WHITE_KING
		else -> error("Invalid piece type: $pieceType")
	}

	BLACK -> when (pieceType) {
		PAWN -> BLACK_PAWN
		KNIGHT -> BLACK_KNIGHT
		BISHOP -> BLACK_BISHOP
		ROOK -> BLACK_ROOK
		QUEEN -> BLACK_QUEEN
		KING -> BLACK_KING
		else -> error("Invalid piece type: $pieceType")
	}

	else -> error("Invalid color: $color")
}

fun pieceColorOf(piece: Piece): Color = when (piece) {
	WHITE_PAWN, WHITE_KNIGHT, WHITE_BISHOP, WHITE_ROOK, WHITE_QUEEN, WHITE_KING -> WHITE
	BLACK_PAWN, BLACK_KNIGHT, BLACK_BISHOP, BLACK_ROOK, BLACK_QUEEN, BLACK_KING -> BLACK
	else -> error("Invalid piece: $piece")
}

fun pieceTypeOf(piece: Piece): PieceType = when (piece) {
	WHITE_PAWN, BLACK_PAWN -> PAWN
	WHITE_KNIGHT, BLACK_KNIGHT -> KNIGHT
	WHITE_BISHOP, BLACK_BISHOP -> BISHOP
	WHITE_ROOK, BLACK_ROOK -> ROOK
	WHITE_QUEEN, BLACK_QUEEN -> QUEEN
	WHITE_KING, BLACK_KING -> KING
	else -> error("Invalid piece: $piece")
}

File: pulse-go/internal/pulse/engine/piece.go
/*
 * Copyright 2013-2024 Phokham Nonava
 *
 * Use of this source code is governed by the MIT license that can be
 * found in the LICENSE file.
 */

package engine

import "fmt"

type Piece = int

const (
	WhitePawn   Piece = 0
	WhiteKnight Piece = 1
	WhiteBishop Piece = 2
	WhiteRook   Piece = 3
	WhiteQueen  Piece = 4
	WhiteKing   Piece = 5
	BlackPawn   Piece = 6
	BlackKnight Piece = 7
	BlackBishop Piece = 8
	BlackRook   Piece = 9
	BlackQueen  Piece = 10
	BlackKing   Piece = 11

	NoPiece Piece = 12
)

var (
	Pieces = [12]Piece{
		WhitePawn, WhiteKnight, WhiteBishop, WhiteRook, WhiteQueen, WhiteKing,
		BlackPawn, BlackKnight, BlackBishop, BlackRook, BlackQueen, BlackKing,
	}
)

func IsValidPiece(pc Piece) bool {
	switch pc {
	case WhitePawn, WhiteKnight, WhiteBishop, WhiteRook, WhiteQueen, WhiteKing, BlackPawn, BlackKnight, BlackBishop, BlackRook, BlackQueen, BlackKing:
		return true
	default:
		return false
	}
}

func PieceOf(col Color, pt PieceType) Piece {
	switch col {
	case White:
		switch pt {
		case Pawn:
			return WhitePawn
		case Knight:
			return WhiteKnight
		case Bishop:
			return WhiteBishop
		case Rook:
			return WhiteRook
		case Queen:
			return WhiteQueen
		case King:
			return WhiteKing
		default:
			panic(fmt.Sprintf("Invalid piece type: %v", pt))
		}
	case Black:
		switch pt {
		case Pawn:
			return BlackPawn
		case Knight:
			return BlackKnight
		case Bishop:
			return BlackBishop
		case Rook:
			return BlackRook
		case Queen:
			return BlackQueen
		case King:
			return BlackKing
		default:
			panic(fmt.Sprintf("Invalid piece type: %v", pt))
		}
	default:
		panic(fmt.Sprintf("Invalid color: %v", col))
	}
}

func PieceColorOf(pc Piece) Color {
	switch pc {
	case WhitePawn, WhiteKnight, WhiteBishop, WhiteRook, WhiteQueen, WhiteKing:
		return White
	case BlackPawn, BlackKnight, BlackBishop, BlackRook, BlackQueen, BlackKing:
		return Black
	default:
		panic(fmt.Sprintf("Invalid piece: %v", pc))
	}
}

func PieceTypeOf(pc Piece) PieceType {
	switch pc {
	case WhitePawn, BlackPawn:
		return Pawn
	case WhiteKnight, BlackKnight:
		return Knight
	case WhiteBishop, BlackBishop:
		return Bishop
	case WhiteRook, BlackRook:
		return Rook
	case WhiteQueen, BlackQueen:
		return Queen
	case WhiteKing, BlackKing:
		return King
	default:
		panic(fmt.Sprintf("Invalid piece: %v", pc))
	}
}
